qiskit/aqua/operators/primitive_ops/primitive_op.py

```diff
@@ -156,7 +156,15 @@ def tensorpower(self, other: int) -> Union[OperatorBase, int]:
         return temp
 
     def compose(self, other: OperatorBase) -> OperatorBase:
-        raise NotImplementedError
+        from ..list_ops.composed_op import ComposedOp
+        if isinstance(other, ComposedOp):
+            comp_with_first = self.compose(other.oplist[0])
+            if not isinstance(comp_with_first, ComposedOp):
+                new_oplist = [comp_with_first] + other.oplist[1:]
+                return ComposedOp(new_oplist, coeff=other.coeff)
+            return ComposedOp([self] + other.oplist, coeff=other.coeff)  # type: ignore
```

Comment:
Can you elaborate on this approach a bit? 

Wouldn't the easiest approach be to do
```
if isinstance(other, ListOp):
   return other.__class__([self.compose(op) for op in other.oplist])
```

Reply:
Hi @Cryosis, 
I think you can not define ```compose``` for ListOp in general.

In ComposedOp, the intepretation of the ```self.oplist``` is that the elements are composed in the sequence.

Hence, if you have PrimitiveOp @ ComposedOp, you are composing PrimitiveOp only with the first operator in the oplist.

This is consistent with the way it is done in case of ComposedOp @ PrimitiveOp.

This is excerpt from ComposedOp:
```python
# Try composing with last element of oplist. We only try
        # this if that last element isn't itself an
        # ComposedOp, so we can tell whether composing the
        # two elements directly worked. If it doesn't,
        # continue to the final return statement below, appending other to the oplist.
        if not isinstance(self.oplist[-1], ComposedOp):
            comp_with_last = self.oplist[-1].compose(other)
            # Attempt successful
            if not isinstance(comp_with_last, ComposedOp):
                new_oplist = self.oplist[0:-1] + [comp_with_last]
                return ComposedOp(new_oplist, coeff=self.coeff)
```

Reply:
It looks like the PR simply implements for `PrimitiveOp` on the LHS the same thing that is already implemented for `PrimitiveOp` on the RHS. And, I agree that the suggestio of @cryois, `return other.__class__([self.compose(op) for op in other.oplist])`, incorrectly composes the `PrimitiveOp` with each element in the list.

Reply:
True -- but this is only the case for `ComposedOp` right? For all others: `ListOp TensoredOp SummedOp` this would be the right behaviour?

Reply:
True -- for `ComposedOp` and `TensoredOp` this wouldn't be the correct behaviour. However for `ListOp` and `SummedOp` this would be right, I think, since you compose with every operator in the list or with all the summands. You're right that `ComposedOp` and `TensoredOp` must be handled separately though

Reply:
Yes, it is adding to PrimitiveOp the already existing feature from ComposedOp.

So should I keep it this way?